src/box/relay.cc

```diff
@@ -188,6 +188,14 @@ struct relay {
 	double txn_lag;
 	/** Relay sync state. */
 	enum relay_state state;
+	/** Whether relay should speed up the next heartbeat dispatch. */
+	bool need_new_vclock_sync;
+	/**
+	 * Whether relay is in process of sending a tx to the remote peer. In
+	 * this case cannot interrupt the transaction stream with any heartbeats
+	 * or other requests.
+	 */
+	bool is_sending_tx;
```

Comment:
Previously `relay_send_heartbeat` was only triggered after a `replication_timeout` has passed after `relay->last_row_time`. So even if it occured while scanning or sending a WAL, it was 99.999% not sent, because `last_row_time` was fresh.
Now when heartbeat is needed, it is sent regardless of `last_row_time`, so it's possible to break a tx being sent by such row. That's why there is this check.

There was a tiny chance of transaction boundary violation with a heartbeat before (if relay stalls after sending one row for too long), now this chance is much higher.